src/Controllers/UserController.js

```diff
@@ -45,14 +49,56 @@ export class UserController extends AdaptableController {
       // TODO: Better error here.
       throw undefined;
     }
-    return this.config.database.update('_User', {
+
+    let query = {username: username, _email_verify_token: token};
+
+    let updateFields = { emailVerified: true, _email_verify_token: {__op: 'Delete'}};
+    if (this.config.allowEmailVerifyTokenToExpire) {
+      updateFields._email_verify_token_expires_at = {__op: 'Delete'};
+    }
+
+    return this.checkEmailTokenValidity(username, token).then((user) => {
+      return this.config.database.update('_User', query, updateFields).then((document) => {
+        if (!document) {
+          throw undefined;
+        }
+        return Promise.resolve(document);
+      });
+
+    });
+  }
+
+  checkEmailTokenValidity(username, token) {
+    // don't check for token validity
+    if (!this.config.allowEmailVerifyTokenToExpire) {
+      return Promise.resolve();
+    }
+
+    return this.config.database.find('_User', {
       username: username,
       _email_verify_token: token
-    }, {emailVerified: true}).then(document => {
-      if (!document) {
+    }, {limit: 1}).then(results => {
```

Comment:
Instead of this, I'd suggest adding a condition to the update that the current time must be before the expiry date, then if that fails you can do a query for the object to find the reason why it failed. That way the happy path requires one less query, and it will be less racy.